### PR TITLE
Render dir

### PR DESCRIPTION
**Pull request by Stevent-fei**

### CODE REVIEW

File: pkg/filesystem/filesystem.go [hunk 1]
```diff
@@ -151,13 +151,25 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
+	renderManifests := fmt.Sprintf(common.RenderManifestsDir, cluster.Name)
 	// TODO scp sdk has change file mod bug
 	initCmd := fmt.Sprintf(RemoteChmod, target)
 	envProcessor := env.NewEnvProcessor(cluster)
 	for _, IP := range ipList {
 		wg.Add(1)
 		go func(ip string) {
 			defer wg.Done()
+			if utils.IsExist(renderEtc) {
+				envProcessor.RenderAll(ip, renderEtc)
```
Comment (bxy4543) [inline pkg/filesystem/filesystem.go:165]: err not return

File: pkg/filesystem/filesystem.go [hunk 2]
```diff
@@ -151,13 +151,34 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
 	initCmd := fmt.Sprintf(RemoteChmod, target)
 	envProcessor := env.NewEnvProcessor(cluster)
 	for _, IP := range ipList {
 		wg.Add(1)
 		go func(ip string) {
 			defer wg.Done()
+			if utils.IsExist(renderEtc) {
+				err := envProcessor.RenderAll(ip, renderEtc)
+				if err != nil {
+					return
```
Comment (bxy4543) [inline pkg/filesystem/filesystem.go:167]: errCh <- err
return

File: pkg/filesystem/filesystem.go [hunk 3]
```diff
@@ -151,13 +151,37 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
+	renderChart := fmt.Sprintf(common.RenderChartsDir, cluster.Name)
+	renderManifests := fmt.Sprintf(common.RenderManifestsDir, cluster.Name)
 	// TODO scp sdk has change file mod bug
 	initCmd := fmt.Sprintf(RemoteChmod, target)
 	envProcessor := env.NewEnvProcessor(cluster)
 	for _, IP := range ipList {
 		wg.Add(1)
 		go func(ip string) {
 			defer wg.Done()
+			if utils.IsExist(renderEtc) {
```
Comment (fanux) [inline pkg/filesystem/filesystem.go:164]: for _,dir = range []{etc,chart, manifests}{
  .....
}

**Comment by fanux** [comment]
https://pkg.go.dev/golang.org/x/sync/errgroup   using error group to handle error

File: pkg/filesystem/filesystem.go [hunk 4]
```diff
@@ -151,13 +153,28 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
+	renderManifests := filepath.Join(common.DefaultMountCloudImageDir(cluster.Name), common.RenderManifestsDir)
 	// TODO scp sdk has change file mod bug
 	initCmd := fmt.Sprintf(RemoteChmod, target)
 	envProcessor := env.NewEnvProcessor(cluster)
 	for _, IP := range ipList {
 		wg.Add(1)
 		go func(ip string) {
 			defer wg.Done()
+			g := new(errgroup.Group)
+			g.Go(func() error {
```
Comment (bxy4543) [inline pkg/filesystem/filesystem.go:167]: Go calls the given function in a new goroutine.
The first call to return a non-nil error cancels the group; its error will be returned by Wait.

'errgroup.group. Go(func() error)' is already a goroutine

File: common/common.go [hunk 5]
```diff
@@ -61,6 +61,10 @@ const (
 	DefaultCloudProvider          = AliCloud
 	ClusterfileName               = "ClusterfileName"
 	CacheID                       = "cacheID"
+	RenderEtcDir                  = "/var/lib/sealer/data/%s/mount/etc"
```
Comment (fanux) [inline common/common.go:64]: Seems you not use those consts

File: pkg/filesystem/filesystem.go [hunk 6]
```diff
@@ -146,37 +148,47 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
-					errCh <- fmt.Errorf("exec init.sh failed %v", err)
+					return fmt.Errorf("exec init.sh failed %v", err)
 				}
 			}
-		}(IP)
+			return err
+		})
+	}
+	if err := g.Wait(); err == nil {
+		fmt.Println(err)
```
Comment (fanux) [inline pkg/filesystem/filesystem.go:190]: You should return the error

File: pkg/filesystem/filesystem.go [hunk 7]
```diff
@@ -146,37 +148,47 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
 			}
-		}(IP)
+			return err
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return err
 	}
-	wg.Wait()
 	return runtime.ReadChanError(errCh)
```
Comment (fanux) [inline pkg/filesystem/filesystem.go:192]: You already use error group, why readChanError again?

File: pkg/filesystem/filesystem.go [hunk 8]
```diff
@@ -146,37 +148,47 @@ func mountRootfs(ipList []string, target string, cluster *v2.Cluster, initFlag b
...
 				if err != nil {
-					errCh <- fmt.Errorf("exec init.sh failed %v", err)
+					return fmt.Errorf("exec init.sh failed %v", err)
 				}
 			}
-		}(IP)
+			return err
+		})
+	}
+	if err := g.Wait(); err != nil {
```
Comment (fanux) [inline pkg/filesystem/filesystem.go:189]: return g.Wait()

**Review by fanux — APPROVED** [review]

**Pull request merged**